File: xcodeproject/SwiftCppTut/Person.swift
//
//  Person.swift
//  SwiftCppTut
//
//  Created by Artur Hellmann on 08.01.17.
//  Copyright © 2017 hellmann. All rights reserved.
//

import Cocoa

class Person: NSObject {
    public let cpp_pointer: UnsafeMutableRawPointer
    
    public var name: String {
        get {
            return String(cString: person_getName(self.cpp_pointer))
        }
        set {
            person_setName(self.cpp_pointer, newValue.cCharArray)
        }
    }
    public var age: Int {
        get {
            return Int(person_getAge(self.cpp_pointer))
        }
        set {
            person_setAge(self.cpp_pointer, Int32(newValue))
        }
    }
    
    init(name: String, age: Int) {
        self.cpp_pointer = person_init(name.cCharArray, Int32(age));
    }
    
    init(withPointer pointer: UnsafeMutableRawPointer) {
        self.cpp_pointer = pointer
        super.init()
    }
    
    deinit {
        person_deconstructor(self.cpp_pointer)
    }
    
}

File: clionproject/Person.cpp
//
// Created by Artur Hellmann on 07.01.17.
//

#include "Person.h"


Person::Person(const string &name, int age) : name(name), age(age) {}

Person::Person() : name("John Doe"), age(29) {}

const string &Person::getName() const {
    return name;
}

void Person::setName(const string &name) {
    Person::name = name;
}

int Person::getAge() const {
    return age;
}

void Person::setAge(int age) {
    Person::age = age;
}

